sotodlib/io/bookbinder.py

```diff
@@ -94,73 +95,156 @@ def next_scan(it):
+    data: np.ndarray
+        HK sample data.
+    finalized: bool
+        True if datas has been processed and finalized.
+    """
+    def __init__(self, instance_id: str, feed: str, field: str):
+        self.instance_id = instance_id
+        self.feed = feed
+        self.field = field
+
+        self.times =[]
+        self.data = []
+        self.finalized = False
+
+    def __len__(self):
+        return len(self.times)
+
+    @property
+    def addr(self):
+        """Returns full address of field"""
+        return f"{self.instance_id}.{self.feed}.{self.field}"
+
     def process_frame(self, frame):
-        if self.name not in frame['block_names']:
+        """Update data based on G3Frame"""
+        address = frame['address']  # "<site>.<instance_id>.feeds.<feed_name>""
+        spl = address.split('.')
+        instance_id, feed = spl[1], spl[3]
+        if instance_id != self.instance_id or feed != self.feed:
             return
-        idx = list(frame['block_names']).index(self.name)
-        block = frame['blocks'][idx]
-
-        self.times.append(np.array(block.times) / core.G3Units.s)
-        for k, v in block.items():
-            if k not in self.data:
-                self.data[k] = []
-            self.data[k].append(v)
-    
-    def finalize(self,drop_duplicates=False):
-        if self.times:
-            self.times = np.hstack(self.times)
-            clean_times, idxs = np.unique(self.times, return_index=True)
-            if len(self.times) != len(clean_times):
-                if not drop_duplicates:
-                    raise DuplicateAncillaryData(
-                        f"HK data from block {self.name} has" 
-                        " duplicate timestamps"
-                    )
-                self.times = self.times[idxs]
-            assert (np.all(np.diff(self.times)>0), f"Times from {self.name} are"
-                            " not increasing")
-            for k, v in self.data.items():
-                self.data[k] = np.hstack(v)[idxs]
-        else:
-            self.times = np.array([], dtype=np.float64)
-            self.data = {}
 
+        for block in frame['blocks']:
+            if self.field not in block:
+                continue
+            self.times.append(np.array(block.times) / core.G3Units.s)
+            self.data.append(block[self.field])
+
+    def finalize(self, drop_duplicates=False):
+        """Finalize data, and store in numpy array"""
+        self.times = np.hstack(self.times, dtype=np.float64)
+        self.data = np.hstack(self.data)
+        self.finalized = True
+
+        # Check for duplicates
+        clean_times, idxs = np.unique(self.times, return_index=True)
+        if len(self.times) != len(clean_times):
+            if not drop_duplicates:
+                raise DuplicateAncillaryData(
+                    f"HK data from {self.addr} has" 
+                    " duplicate timestamps"
+                )
+            self.times = self.times[idxs]
```

Comment:
Don't you need to take subset of self.data here too?

Reply:
yes, fixed in a soon to update push